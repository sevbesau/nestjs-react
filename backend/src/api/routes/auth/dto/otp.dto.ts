import { validateOtpRequestSchema } from '@common/schemas';
import { createZodDto } from 'nestjs-zod';

export class ValidateOtpRequestDto extends createZodDto(
  validateOtpRequestSchema,
) {}
