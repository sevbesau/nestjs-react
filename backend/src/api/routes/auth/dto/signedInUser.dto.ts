import { signedInUserResponseSchema } from '@common/schemas';
import { createZodDto } from 'nestjs-zod';

export class SignedInUserResponseDto extends createZodDto(
  signedInUserResponseSchema,
) {}
