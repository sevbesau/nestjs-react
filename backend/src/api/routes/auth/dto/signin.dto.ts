import {
  signInRequestSchema,
  signInResponseSchema,
  userZodSchema,
} from '@common/schemas';
import { createZodDto } from 'nestjs-zod';
import { z } from 'zod';

export const signInRequestZodSchema = z.discriminatedUnion('strategy', [
  z
    .object({
      strategy: z.literal('OTP'),
    })
    .merge(
      userZodSchema.pick({
        email: true,
      }),
    ),
  z
    .object({
      strategy: z.literal('PASSWORD'),
    })
    .merge(
      userZodSchema.pick({
        email: true,
        // TODO: password
      }),
    ),
]);

export class SignInRequestDto extends createZodDto(signInRequestSchema) {}

export class SignInResponseDto extends createZodDto(signInResponseSchema) {}
