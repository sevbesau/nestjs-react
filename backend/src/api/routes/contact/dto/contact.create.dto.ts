import { contactRequestSchema } from '@common/schemas';
import { createZodDto } from 'nestjs-zod';

export class ContactRequestDto extends createZodDto(contactRequestSchema) {}

export const createContactSchema = contactRequestSchema.omit({
  captchaResponse: true,
});
