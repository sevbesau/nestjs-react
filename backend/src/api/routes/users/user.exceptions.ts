import { NotFoundException } from '@nestjs/common';

export class UserNotFoundException extends NotFoundException {
  constructor() {
    super('user not found');
  }
}
