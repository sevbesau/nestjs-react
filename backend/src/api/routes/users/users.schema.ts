import { TAddress, TOtp, TUser, UserRole } from '@common/schemas';
import { Address } from '@lib/common/address.schema';
import { BaseSchema, BaseSchemaOptions } from '@lib/common/base.schema';
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document, HydratedDocument } from 'mongoose';

export class Otp extends Document implements TOtp {
  @Prop({ required: true })
  password: string;
  @Prop({ required: true })
  expiresAt: number;
}

@Schema(BaseSchemaOptions)
export class User extends BaseSchema implements TUser {
  @Prop({ required: true, type: Array<Otp>, default: [] })
  otps: TOtp[];

  @Prop({ required: true, type: Address })
  address: TAddress;

  @Prop({ required: true })
  firstName: string;

  @Prop({ required: true })
  lastName: string;

  @Prop({ required: true, unique: true })
  email: string;

  @Prop({ required: true })
  phone: string;

  @Prop({ required: true, type: Array<string>, default: [UserRole.USER] })
  roles: UserRole[];

  @Prop({ required: true, default: false })
  verified: boolean;

  @Prop({ required: true, default: false })
  onboarded: boolean;

  @Prop({ required: true, default: false })
  blocked: boolean;

  @Prop({ required: true, default: 0 })
  tokenVersion: number;
}

export type UserDocument = HydratedDocument<User>;

export const UserSchema = SchemaFactory.createForClass(User);
