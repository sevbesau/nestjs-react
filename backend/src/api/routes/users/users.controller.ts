import { GetSession, Session } from '@lib/common/session';
import { Body, Controller, Patch } from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { ZodSerializerDto } from 'nestjs-zod';

import { SanitizedUserDto } from './dto/user.sanitized.dto';
import { UpdateUserDto } from './dto/user.update.dto';
import { UsersService } from './users.service';

@Controller()
@ApiBearerAuth()
@ApiTags(UsersController.BasePath)
export class UsersController {
  static readonly BasePath = 'users';

  constructor(private readonly usersService: UsersService) {}

  @Patch()
  @ApiOperation({ summary: 'Updates the logged in users profile.' })
  @ZodSerializerDto(SanitizedUserDto)
  @ApiOkResponse({ type: SanitizedUserDto })
  updateLoggedInUser(
    @GetSession() session: Session,
    @Body() updateUserDto: UpdateUserDto,
  ) {
    return this.usersService.updateById(session.sub, updateUserDto);
  }
}
