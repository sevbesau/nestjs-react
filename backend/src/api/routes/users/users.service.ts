import { TOtp, UserRole } from '@common/schemas';
import { Injectable } from '@nestjs/common';

import { CreateUserDto } from './dto/user.create.dto';
import { UpdateUserDto } from './dto/user.update.dto';
import { UsersRepository } from './users.repository';
import { User } from './users.schema';

@Injectable()
export class UsersService {
  constructor(private readonly usersRepository: UsersRepository) {}

  async createOne(createUserDto: CreateUserDto): Promise<User> {
    return this.usersRepository.createOne(createUserDto);
  }

  async findOneByEmailOrCreateOne(createUserDto: CreateUserDto) {
    try {
      return await this.usersRepository.findOneByEmailOrFail(
        createUserDto.email,
      );
    } catch (error) {
      return this.usersRepository.createOne(createUserDto);
    }
  }

  async findById(userId: string) {
    return this.usersRepository.findOneByIdOrFail(userId);
  }

  async findByEmail(email: string) {
    return this.usersRepository.findOneByEmailOrFail(email);
  }

  async isDuplicateEmail(email: string): Promise<boolean> {
    try {
      await this.findByEmail(email);
      return true;
    } catch {
      return false;
    }
  }

  async addRoleById(userId: string, role: UserRole) {
    return this.usersRepository.addRoleByIdOrFail(userId, role);
  }

  async addOtpById(userId: string, otp: TOtp) {
    return this.usersRepository.addOtpByIdOrFail(userId, otp);
  }

  async clearOtpsById(userId: string) {
    return this.usersRepository.clearOtpsByIdOrFail(userId);
  }

  async verifyById(userId: string) {
    return this.usersRepository.verifyByIdOrFail(userId);
  }

  async updateById(userId: string, updateUserDto: UpdateUserDto) {
    return this.usersRepository.overwriteOneByIdOrFail(userId, updateUserDto);
  }

  async incrementTokenVersionById(userId: string) {
    return this.usersRepository.incrementTokenVersionByIdOrFail(userId);
  }
}
