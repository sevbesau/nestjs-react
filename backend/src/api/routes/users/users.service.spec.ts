import { UserRole } from '@common/schemas';
import { faker } from '@faker-js/faker';
import { MyConfigModule } from '@lib/config/config.module';
import { DataBaseModule } from '@lib/database/database.module';
import { userDocFactory, userFactory } from '@lib/test/factories';
import { MongooseModule } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';
import { Model } from 'mongoose';

import { CreateUserDto } from './dto/user.create.dto';
import { UserNotFoundException } from './user.exceptions';
import { User, UserSchema } from './users.schema';
import { UsersService } from './users.service';

describe('UsersService', () => {
  let module: TestingModule;
  let usersService: UsersService;

  beforeEach(async () => {
    module = await Test.createTestingModule({
      imports: [
        MyConfigModule,
        DataBaseModule,
        MongooseModule.forFeature([{ name: User.name, schema: UserSchema }]),
      ],
      providers: [UsersService],
    }).compile();

    usersService = module.get<UsersService>(UsersService);
  });

  afterEach(async () => {
    await module.close();
  });

  it('should be defined', () => {
    expect(usersService).toBeDefined();
  });

  describe('createOne', () => {
    it('should be defined', () => {
      expect(usersService.createOne).toBeDefined();
    });

    it('should insert a new user into the database', async () => {
      const userDto = userFactory.build();
      const result = await usersService.createOne(userDto);
      expect(result).toBeDefined();
      expect(result).toMatchObject<Model<User>>;
      const createdUser = await usersService.findById(result._id);
      expect(createdUser).toBeDefined();
    });

    it('should reject invalid input', async () => {
      const userDto: CreateUserDto = {
        firstName: '',
        lastName: '',
        email: '',
        phone: '',
        address: {
          street: '',
          city: '',
          zip: '',
        },
      };
      await expect(
        usersService.createOne(userDto),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('findById', () => {
    it('should be defined', () => {
      expect(usersService.findById).toBeDefined();
    });

    it('should find a user in the database by id', async () => {
      const user = await usersService.createOne(userFactory.build());
      const foundUser = await usersService.findById(user._id);
      expect(foundUser).toBeDefined();
    });

    it('should throw when the user is not found', async () => {
      await expect(
        usersService.findById(faker.database.mongodbObjectId()),
      ).rejects.toThrow(UserNotFoundException);
    });
  });

  describe('findByEmail', () => {
    it('should be defined', () => {
      expect(usersService.findByEmail).toBeDefined();
    });

    it('should find a user in the database by email', async () => {
      const user = await usersService.createOne(userFactory.build());
      const foundUser = await usersService.findByEmail(user.email);
      expect(foundUser).toBeDefined();
    });

    it('should throw when the user is not found', async () => {
      await expect(
        usersService.findById(faker.database.mongodbObjectId()),
      ).rejects.toThrow(UserNotFoundException);
    });
  });

  describe('isDuplicateEmail', () => {
    it('should be defined', () => {
      expect(usersService.isDuplicateEmail).toBeDefined();
    });

    it('should return true if a user exists with a given email', async () => {
      const user = await usersService.createOne(userFactory.build());
      const isDuplicateEmail = await usersService.isDuplicateEmail(user.email);
      expect(isDuplicateEmail).toBe(true);
    });

    it('should return false if no user exists with a given email', async () => {
      const { email } = userFactory.build();
      const isDuplicateEmail = await usersService.isDuplicateEmail(email);
      expect(isDuplicateEmail).toBe(false);
    });
  });

  describe('addRoleById', () => {
    it('should be defined', () => {
      expect(usersService.addRoleById).toBeDefined();
    });

    it('should add a new role to the users roles', async () => {
      const user = await usersService.createOne(userFactory.build());
      const updatedUser = await usersService.addRoleById(
        user._id,
        UserRole.GOD,
      );

      expect(updatedUser.roles.length).toEqual(user.roles.length + 1);
      expect(updatedUser.roles).toEqual([...user.roles, UserRole.GOD]);
    });

    it('should not add duplicate roles', async () => {
      const user = await usersService.createOne(userFactory.build());
      const updatedUser = await usersService.addRoleById(
        user._id,
        UserRole.RECIPIENT,
      );
      expect(updatedUser.roles).toEqual(user.roles);
    });

    it('should throw when the user is not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(
        usersService.addRoleById(mockUser._id, UserRole.GOD),
      ).rejects.toThrow(UserNotFoundException);
    });
  });

  describe('addOtpById', () => {
    const mockOtp = {
      password: 'password',
      expiresAt: Date.now(),
    };

    it('should be defined', () => {
      expect(usersService.addOtpById).toBeDefined();
    });

    it('should add an otp to a user and return the updated doc', async () => {
      const user = await usersService.createOne(userFactory.build());
      const updatedUser = await usersService.addOtpById(user._id, mockOtp);
      expect(updatedUser?.otps).toMatchObject([mockOtp]);
    });

    it('should throw when the user is not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(
        usersService.addOtpById(mockUser._id, mockOtp),
      ).rejects.toThrow(UserNotFoundException);
    });
  });

  describe('clearOtpsById', () => {
    it('should be defined', () => {
      expect(usersService.clearOtpsById).toBeDefined();
    });

    it('should remove all otps from a user and return the updated doc', async () => {
      const user = await usersService.createOne(userFactory.build());
      const mockOtp = {
        password: 'password',
        expiresAt: Date.now(),
      };
      await Promise.all([
        usersService.addOtpById(user._id, mockOtp),
        usersService.addOtpById(user._id, mockOtp),
      ]);

      const updatedUser = await usersService.clearOtpsById(user._id);

      expect(updatedUser?.otps).toMatchObject([]);
    });

    it('should throw when the user if not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(usersService.clearOtpsById(mockUser._id)).rejects.toThrow(
        UserNotFoundException,
      );
    });
  });

  describe('verifyById', () => {
    it('should be defined', () => {
      expect(usersService.verifyById).toBeDefined();
    });

    it('should verify the user and return the updated doc', async () => {
      const user = await usersService.createOne(userFactory.build());
      expect(user.verified).toBe(false);

      const updatedUser = await usersService.verifyById(user._id);

      expect(updatedUser._id).toEqual(user._id);
      expect(updatedUser.verified).toBe(true);
    });

    it('should throw when the user is not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(usersService.verifyById(mockUser._id)).rejects.toThrow(
        UserNotFoundException,
      );
    });
  });

  describe('updateById', () => {
    it('should be defined', () => {
      expect(usersService.updateById).toBeDefined();
    });

    it('should update a user and return the updated doc', async () => {
      const user = await usersService.createOne(userFactory.build());
      const newFirstName = 'other';
      const newLastName = 'name';
      const updatedUser = await usersService.updateById(user._id, {
        firstName: newFirstName,
        lastName: newLastName,
      });
      expect(updatedUser).toBeDefined();
      expect(updatedUser.firstName).toBe(newFirstName);
      expect(updatedUser.lastName).toBe(newLastName);
    });

    it('should throw when the user is not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(
        usersService.updateById(mockUser._id, mockUser),
      ).rejects.toThrow(UserNotFoundException);
    });
  });

  describe('incrementTokenVersionById', () => {
    it('should be defined', () => {
      expect(usersService.incrementTokenVersionById).toBeDefined();
    });

    it('should increment the tokenversion of a user and return the updated doc', async () => {
      const user = await usersService.createOne(userFactory.build());
      const updatedUser = await usersService.incrementTokenVersionById(
        user._id,
      );
      expect(updatedUser?.tokenVersion).toBe(user.tokenVersion + 1);
    });

    it('should throw when the user is not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(
        usersService.incrementTokenVersionById(mockUser._id),
      ).rejects.toThrow(UserNotFoundException);
    });
  });

  describe('deleteOneById', () => {
    it('should be defined', () => {
      expect(usersService.deleteOneById).toBeDefined();
    });

    it('should increment the tokenversion of a user and return the updated doc', async () => {
      const user = await usersService.createOne(userFactory.build());
      const deletedUser = await usersService.deleteOneById(user._id);
      expect(deletedUser?._id).toEqual(user._id);
      await expect(
        usersService.findById(deletedUser?._id as string),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw when the user is not found', async () => {
      const mockUser = userDocFactory.build();
      await expect(usersService.deleteOneById(mockUser._id)).rejects.toThrow(
        UserNotFoundException,
      );
    });
  });
});
