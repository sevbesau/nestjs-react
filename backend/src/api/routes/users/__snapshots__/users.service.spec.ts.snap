// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService createOne should reject invalid input 1`] = `
"[
  {
    "validation": "email",
    "code": "invalid_string",
    "message": "Invalid email",
    "path": [
      "email"
    ]
  },
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "phone"
    ]
  },
  {
    "code": "too_small",
    "minimum": 2,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 2 character(s)",
    "path": [
      "address",
      "street"
    ]
  },
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "address",
      "street"
    ]
  },
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "address",
      "zip"
    ]
  },
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "address",
      "zip"
    ]
  },
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "address",
      "city"
    ]
  },
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "address",
      "city"
    ]
  }
]"
`;

exports[`UsersService deleteOneById should increment the tokenversion of a user and return the updated doc 1`] = `"user not found"`;
